widget/numerical_entry.go

```diff
@@ -72,12 +177,253 @@ func (e *NumericalEntry) Keyboard() mobile.KeyboardType {
 	return mobile.NumberKeyboard
 }
 
-func (e *NumericalEntry) isNumber(content string) bool {
-	if e.AllowFloat {
-		_, err := strconv.ParseFloat(content, 64)
-		return err == nil
+// Unbind disconnects the entry from the bound data.
+// This will remove the data updates and allow the entry to be used independently.
+func (e *NumericalEntry) Unbind() {
+	e.binder.Unbind()
+}
+
+// ValidateText checks if the entered text is a valid numerical input
+// according to the system locale.
+func (e *NumericalEntry) ValidateText(text string) error {
+	if len(text) == 0 {
+		return nil
+	}
+	runes := []rune(text)
+	if !e.AllowNegative && runes[0] == e.minus {
+		return errors.New("negative numbers are not allowed")
+	}
+	if !e.AllowFloat && strings.Contains(text, string(e.radixSep)) {
+		return errors.New("floating point numbers are not allowed")
+	}
+	radixCount := 0
+	for i, r := range runes {
+		if unicode.IsDigit(r) {
+			continue
+		}
+		if r != e.minus && r != e.radixSep && r != e.thouSep {
+			return fmt.Errorf("invalid character: %q", r)
+		}
+		if r == e.minus && i != 0 {
+			return errors.New("minus must be the first character")
+		}
+		if r == e.radixSep {
+			radixCount++
+			if radixCount > 1 {
+				return errors.New("only one radix separator is allowed")
+			}
+			if i > 0 && runes[i-1] == e.thouSep {
+				return errors.New("thousand separator cannot immediately precede radix separator")
+			}
+		}
+		if r == e.thouSep {
+			if i == 0 {
+				return errors.New("thousand separator cannot be the first character")
+			} else if i == 1 && runes[0] == e.minus {
+				return errors.New("thousand separator cannot be the first character after minus")
+			} else if runes[i-1] == e.radixSep {
+				return errors.New("thousand separator cannot be immediately after radix separator")
+			} else if runes[i-1] == e.thouSep {
+				return errors.New("thousand separator cannot be immediately after thousand separator")
+			}
+
+		}
+	}
+	return nil
+}
+
+// getRuneForLocale checks if a rune is valid for the entry,
+// and returns the correct rune for the locale.
+func (e *NumericalEntry) getRuneForLocale(r rune) (rune, bool) {
+	if unicode.IsDigit(r) {
+		return r, true
+	}
+
+	switch r {
+	case '-': // hyphen - minus
```

Comment:
I don't think we need comments to explain a visible rune.

p.s. all the cases that are a fall through could be on one case, shortening this a lot.

Reply:
cases combined without fallthrough comment and character comments removed.